modules/private.php

```diff
@@ -0,0 +1,256 @@
+ * Module Tags: Private
+ * Feature: Traffic
+ * Additional Search Queries: private, sandbox, launch, unlaunched
+ *
+ * @package Jetpack
+ */
+
+class Jetpack_Private {
+	static function init() {
+		add_action( 'parse_request', array( __CLASS__, 'privatize_blog' ), 100 );
+		add_action( 'login_init', array( __CLASS__, 'privatize_blog_maybe_mask_blog_name' ) );
+		add_filter( 'preprocess_comment', array( __CLASS__, 'privatize_blog_comments' ) );
+		add_action( 'blog_privacy_selector', array( __CLASS__, 'privatize_blog_priv_selector' ) );
+		add_filter( 'robots_txt', array( __CLASS__, 'private_robots_txt' ) );
+		add_action( 'wp_head', array( __CLASS__, 'private_no_pinning' ) );
+		add_action( 'check_ajax_referer', array( __CLASS__, 'private_blog_ajax_nonce_check' ), 9, 2 );
+		add_action( 'rest_api_init', array( __CLASS__, 'disable_rest_api' ) );
+		add_filter( 'option_jetpack_active_modules', array( __CLASS__, 'module_override' ) );
+		add_action( 'update_option_blog_public', array( __CLASS__, 'private_update_option_blog_public' ) );
+		add_action( 'update_right_now_text', array( __CLASS__, 'add_private_dashboard_glance_items' ) );
+	}
+
+	/**
+	 * Returns the private site template for private blogs
+	 *
+	 * @param object $wp Current WordPress environment instance (passed by reference).
+	 */
+	static function privatize_blog( $wp ) {
+		global $pagenow, $current_user, $wpdb;
+
+		if ( 'wp-login.php' === $pagenow ) {
+			return;
+		}
+
+		if ( defined( 'WP_CLI' ) && WP_CLI ) {
+			return;
+		}
+
+		// Serve robots.txt for private blogs.
+		if ( is_object( $wp ) && ! empty( $wp->query_vars['robots'] ) ) {
+			return;
+		}
+
+		if ( $current_user && ( is_super_admin() || Jetpack_Private::is_private_blog_user( $wpdb->blogid, $current_user ) ) ) {
+			return;
+		}
+
+		if ( file_exists( JETPACK__PLUGIN_DIR . '/modules/private/private.php' ) ) {
+			include JETPACK__PLUGIN_DIR . '/modules/private/private.php';
+		} else {
+			esc_html_e( 'This site is private.', 'jetpack' );
+		}
+
+		exit;
+	}
+
+	/**
+	 * Does not check whether the blog is private. Accepts blog and user in various types.
+	 * Returns true for super admins; if you don't want that, use is_really_private_blog_user.
+	 *
+	 * @param int   $blog Current WordPress blod id.
+	 * @param Mixed $user Current WordPress user or user_id.
+	 */
+	static function is_private_blog_user( $blog, $user ) {
+		global $wpdb;
+
+		if ( ! is_object( $user ) ) {
+			$user = new WP_User( $user );
+		}
+
+		if ( ! $user->ID ) {
+			return false;
+		}
+
+		if ( is_numeric( $blog ) ) {
+			$blog_id = intval( $blog );
+		} elseif ( is_object( $blog ) ) {
+			$blog_id = $blog->blog_id;
+		} elseif ( is_string( $blog ) ) {
+			$blog    = get_blog_info( $blog, '/', 1 );
+			$blog_id = $blog->blog_id;
+		} else {
+			$blog_id = $wpdb->blogid;
+		}
+
+		// check if the user has read permissions.
+		$the_user = wp_clone( $user );
+		$the_user->for_site( $blog_id );
+		return $the_user->has_cap( 'read' );
```

Comment:
I think it may be worth adding a filter here. I can imagine that some users will ask us if the feature can be used to block other user caps instead of just `read`.

When adding a filter, you can follow this example:
https://github.com/Automattic/jetpack/blob/6.8/modules/sitemaps/sitemaps.php/#L407-L418

Reply:
Done 👍